Switch to Play 3 and go all in Pekko

ioannakok opened the pull request:
Co-authored-by: @mkurz, @kelvin-chappell @mchv @mxdvl   

Re-raising @mkurz's PR:
* https://github.com/guardian/frontend/pull/26677 

It has to be raised by someone in The Guardian organisation for the build to run.

## What does this change?
* Switches to Pekko and Play 3
* Removes Jackson dependencies. Play 3 upgraded to Jackson 2.14. No need to upgrade explicitly anymore. See [comment](https://github.com/guardian/frontend/pull/26677#discussion_r1385208651).
* Upgrades the following dependencies and plugin to their latest versions: 
  - `scalatestplus-play` 
  - `mockito-4-11` 
  - `io.lemonlabs:scala-uri`
  - `sbt-buildinfo`
* Updates Scala version to latest `2.13.13`
* Removes a fruitless filtering in `PressedContent` https://github.com/guardian/frontend/commit/4f55a6987cfe5400995568c29d9e9634ac79cb58

### Previous PRs preparing for this upgrade:
* https://github.com/guardian/frontend/pull/27021
* https://github.com/guardian/frontend/pull/26983
* https://github.com/guardian/frontend/pull/26986
* https://github.com/guardian/frontend/pull/26796
* https://github.com/guardian/frontend/pull/26806
* https://github.com/guardian/frontend/pull/26783
* https://github.com/guardian/frontend/pull/27030
* https://github.com/guardian/frontend/pull/27031
* https://github.com/guardian/frontend/pull/27032

## Checklist

- [x] Tested locally, and on CODE if necessary

ioannakok commented:
@mkurz 👋 from here as well. I have pushed a couple of commits.

We are now unfortunately getting a binary incompatibility error: 

```
[error] (common / update) found version conflict(s) in library dependencies; some are suspected to be binary incompatible:
[error]
[error] 	* org.scala-lang.modules:scala-xml_2.13:2.2.0 (early-semver) is selected over {1.2.0, 1.3.0}
[error] 	    +- org.playframework:play-ws-standalone-xml_2.13:3.0.0 (depends on 2.2.0)
[error] 	    +- org.playframework.twirl:twirl-api_2.13:2.0.1       (depends on 2.2.0)
[error] 	    +- net.liftweb:lift-json_2.13:3.5.0                   (depends on 1.3.0)
[error] 	    +- com.typesafe.play:twirl-api_2.13:1.4.2             (depends on 1.2.0)
```

`lift-json` and `com.typesafe.play:twirl-api` are both coming from libraries that we own but it looks like this is going to be complicated. I will try to summarise my findings here:

### lift-json 

```
+-com.gu.identity:identity-model_2.13:3.255 
   | +-net.liftweb:lift-json_2.13:3.4.3
   | | +-org.scala-lang.modules:scala-xml_2.13:1.3.0
```

Work needs to be done to go to a >= 4.10 identity version which uses the latest `lift-json:3.5.0`. 4.10 has breaking changes. But even when we do that `lift-json:3.5.0` still uses https://index.scala-lang.org/lift/framework/artifacts/lift-json/3.5.0 `org.scala-lang.modules:scala-xml:1.3.0`. 


### com.typesafe.play:twirl-api

```
+-com.gu:atom-renderer_2.13:1.2.0 [S]
   | +-org.http4s:http4s-twirl_2.13:0.21.0-M5	
   | | +-com.typesafe.play:twirl-api_2.13:1.4.2

```

This is actually a deprecated repo but we sometimes have to do upgrades if it blocks other pieces of work, Play 3 is a good example. We will have to either upgrade `org.http4s:http4s-twirl` there or completely remove `atom-renderer` dependency from `frontend`. I'm currently looking if it's possible to do the latter. Will give an update here once I know more!

mkurz commented:
@ioannakok scala-xml 1.x and 2.x are pretty much compatible, please see https://github.com/scala/scala-xml/discussions/605#discussioncomment-2828193, so you should be safe to just exclude it from the dependencies that pull in the old 1.x version.

I checked out this PR and eventually got it compiling. I pushed the commits to my [`switch-to-pekko-and-play3-mkurz`](https://github.com/mkurz/guardian-frontend/tree/switch-to-pekko-and-play3-mkurz) branch:
* You need to switch the package from `org.fluentlenium` to `io.fluentlenium`: https://github.com/mkurz/guardian-frontend/commit/780ddbb898301233c9fe03b827b517f2126cd206
* You need to fix a deprecation: https://github.com/mkurz/guardian-frontend/commit/a436cbe61562ee9a326a71a683e8f240a0ab397d
* And fix another compiler error regarding implicits: https://github.com/mkurz/guardian-frontend/commit/7be6b8371d340c9a594c7a5975e0d6ce560bcda2

You can cherry-pick those commits if you like.

And now to make things compile I just excluded some libraries:
- https://github.com/mkurz/guardian-frontend/commit/c7ce14908150caacac21411b3a4ee38da54c486a

like the old scala-xml dependencies. Also some libraries, the identity ones, still pull in `com.typesafe.play` packages, which means those identity libraries haven't been upgraded to Play 3 yet (?). Also I don't know under which repo they would be hosted: https://github.com/orgs/guardian/repositories?q=identity&type=all&language=&sort= (?)

Liked said, I am pretty sure you could even go in production by having the old scala-xml excluded, however you should take care of the pre Play 3.0 dependencies... Also I saw that you still pull in `com.typesafe.play:play-json` somewhere... There aren't compiler errors, but you should see to upgraded it in the affected dependencies...

Anyway, if you like, you could (at least temporary) cherry-pick that last commit of mine as well, just to make CI run the tests and see if something else needs to be fixed, so you at least are not blocked now.

ioannakok commented:
@mkurz this is great, thanks so much for your help! I've also added a few commits:

* Bumped `io.lemonlabs:scala-uri` to `4.0.3`. This fixes quite a few of [failing tests](https://github.com/guardian/frontend/actions/runs/7184605032/job/19566759418#step:9:1240): https://github.com/guardian/frontend/pull/26755/commits/3ec7c20e08d1f55a4790fa52819f190a338abdc8
* Bumped `enumeratum-play-json` to `1.8.0`. [Uses](https://index.scala-lang.org/lloydmeta/enumeratum/artifacts/enumeratum-play-json/1.8.0) `org.playframework::play-json` https://github.com/guardian/frontend/pull/26755/commits/e909ce3240ecfbb354add23e218ae47b3520fae5
* Temporarily commented out three test suites which hang forever and prevent the build from completing: https://github.com/guardian/frontend/pull/26755/commits/cde9a285f542a885744bf98681303f1a95dc3d04. I would like to see the rest of the failing tests.
* As per a suggestion in https://github.com/bizzabo/play-json-extensions/issues/94 I tried to exclude `com.typesafe.play` from `ai.x:play-json-extensions` but was unsuccessful. Tried both ways that are commented out in the commit https://github.com/guardian/frontend/pull/26755/commits/1c1306e36f5391ece3c26d89252f915f6f2f1d13. I thought this would be enough since you've added `org.playframework:play-json` in https://github.com/guardian/frontend/commit/8e5eecc32e500930fbf289fe9ed04408b8c5b426 but it still pulls in the `com.typesafe.play` one:

```
+-ai.x:play-json-extensions_2.13:0.42.0 [S]
| | +-com.typesafe.play:play-json_2.13:2.8.1 (evicted by: 2.8.2)
| | +-com.typesafe.play:play-json_2.13:2.8.2 [S]
```
Let me know if I'm misunderstanding anything here.
* The identity libraries are hosted under a private repo but will have to check with @guardian/identity devs whether there is a plan to update those to Play 3 any soon. 

Thanks again for your help.

mkurz commented:
> * I tried to exclude `com.typesafe.play` from `ai.x:play-json-extensions` but was unsuccessful. Tried both ways that are commented out in the commit [1c1306e](https://github.com/guardian/frontend/commit/1c1306e36f5391ece3c26d89252f915f6f2f1d13). I thought this would be enough since you've added `org.playframework:play-json` in [8e5eecc](https://github.com/guardian/frontend/commit/8e5eecc32e500930fbf289fe9ed04408b8c5b426) but it still pulls in the `com.typesafe.play` one:

For me it is working. Did you run `reload` within sbt before you tried again? Maybe you need to run `clean` within sbt or quit/restart sbt so changes take effect?
This is what I am using: https://github.com/mkurz/guardian-frontend/commit/ee7bafc1eeb576d748afeef056a4a9af0401f9b7

Assuming `ai.x:play-json-extensions` is "fixed", the only dependencies left pulling in the old play-json are these:

<kbd><img width="600" src="https://github.com/guardian/frontend/assets/644927/e1361650-3531-4c93-8f5f-ce6f6bc4495e" /></kbd>

* For facia-json I opened a pull request (https://github.com/guardian/facia-scala-client/pull/298) and made a note in my branch: https://github.com/mkurz/guardian-frontend/commit/ac37835af7785d01b72c74867d64b0873905ed81
* It seems targeting-client is a private repo, at least I [can not find it](https://github.com/orgs/guardian/repositories?q=targeting-client&type=all&language=&sort=). According to on of its [pom files](https://repo1.maven.org/maven2/com/gu/targeting-client_2.13/1.1.4/targeting-client_2.13-1.1.4.pom) on maven central it should be hosted here: https://github.com/guardian/targeting-client Added a note also: https://github.com/mkurz/guardian-frontend/commit/5c1cfb8725d62002110abedc5f7b6a3db23291fe

mkurz commented on project/Dependencies.scala:
```diff
@@ -33,7 +33,8 @@ object Dependencies {
   val contentApiClient = "com.gu" %% "content-api-client" % capiVersion
   val dfpAxis = "com.google.api-ads" % "dfp-axis" % "5.2.0"
-  val faciaFapiScalaClient = "com.gu" %% "fapi-client-play28" % faciaVersion
+  val faciaFapiScalaClient = "com.gu" %% "fapi-client-play28" % faciaVersion // TODO: should be `fapi-client-play30`
```
```suggestion
val faciaFapiScalaClient = "com.gu" %% "fapi-client-play30" % faciaVersion
```

?

mkurz replied:
`faciaVersion` needs to be set to 4.0.10 I guess

mkurz commented on project/plugins.sbt:
```diff
@@ -7,38 +7,18 @@ libraryDependencies ++= Seq(
+resolvers ++= Resolver.sonatypeOssRepos("releases")
 
-addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.9.1")
+addSbtPlugin("org.playframework" % "sbt-plugin" % "3.0.0")
```
```suggestion
addSbtPlugin("org.playframework" % "sbt-plugin" % "3.0.1")
```

ioannakok commented on project/Dependencies.scala:
```diff
@@ -89,49 +96,22 @@ object Dependencies {
 
   val kinesisLogbackAppender = "com.gu" % "kinesis-logback-appender" % "1.4.0"
-  val targetingClient = "com.gu" %% "targeting-client" % "1.1.4"
+  val targetingClient = "com.gu" %% "targeting-client" % "1.1.4" // TODO: Needs to be upgraded so it pulls in play-json with groupid org.playframework
```
Yes, we need to fix this 👍

mkurz commented:
IMHO still relevant.

mkurz commented:
@mchv latest Play version is 3.0.2 ([release notes](https://github.com/playframework/playframework/releases/tag/3.0.2)), same as play-json ([release notes](https://github.com/playframework/play-json/releases/tag/3.0.2))

mchv commented:
> @mchv latest Play version is 3.0.2 ([release notes](https://github.com/playframework/playframework/releases/tag/3.0.2)), same as play-json ([release notes](https://github.com/playframework/play-json/releases/tag/3.0.2))

Yes, I have updated in a further commit.

mchv commented:
@ioannakok I have rebased against `main` and added a few commits (`sbt` update, `3.0.2` update, update targeting client, etc.) and this now build, but we need to uncomment remaining ignored test and investigate any new failure.

mchv commented on project/ProjectSettings.scala:
```diff
@@ -77,7 +77,6 @@ object ProjectSettings {
     // These settings are needed for forking, which in turn is needed for concurrent restrictions.
     Test / javaOptions += "-DAPP_SECRET=this_is_not_a_real_secret_just_for_tests",
     Test / javaOptions += "-Xmx2048M",
-    Test / javaOptions += "-XX:+UseConcMarkSweepGC",
```
This is not needed anymore, the Concurrent Mark and Sweep (CMS) has been deprecated and we should not use it anymore, see https://openjdk.org/jeps/291.

mxdvl replied:
But does it need to be changed as part of this PR, or should it be a follow-up?

ioannakok commented on project/Dependencies.scala:
```diff
@@ -50,8 +49,14 @@ object Dependencies {
+//      ExclusionRule("com.typesafe.play", "play_2.13"),
+//      ExclusionRule("com.typesafe.play", "twirl-api_2.13"),
+//      ExclusionRule("com.typesafe.play", "twirl-compiler_2.13"),
+//      ExclusionRule("com.typesafe.play", "twirl-parser_2.13"),
```
@mkurz I commented these exclusions and I can't see any `com.typesafe.play` dependency being pulled in from anywhere? Does this mean this is good to go? Same for [lines 108-110](https://github.com/guardian/frontend/pull/26755/files#diff-ad2642dc77b3679e4ad2c7d3b2f59a2dcf48c1bf5b93a558a4fa44828315a34cR108-R110)

mkurz replied:
Please wait with merging, there is still at least `com.typesafe.play:twirl-api_2.13:1.4.2` pulled in, which should not be the case.
I will investigate asap.

ioannakok replied:
Spoke too soon, sorry! I think the things we still need to fix are:

1. As you say: `com.typesafe.play:twirl-api_2.13:1.4.2`. This is coming from [`atom-renderer`](https://github.com/guardian/atom-renderer). It's a legacy repo we have. I can have a look, please feel free to contribute
<img width="479" alt="image" src="https://github.com/guardian/frontend/assets/19683595/9bfbbf96-6f67-4e0c-bbe7-57b9b74f586d">

2. identity project also pulls `com.typesafe.play` dependencies. We need to fix this. This is a Guardian private repo so you can't contribute but I can speak with the team and get back to you.

Also, I forgot to comment out the exclusion for `play-json-extensions` [here](https://github.com/guardian/frontend/pull/26755/files#diff-ad2642dc77b3679e4ad2c7d3b2f59a2dcf48c1bf5b93a558a4fa44828315a34cR103-R104) but just discovered we're not using this anywhere so raising a separate PR to remove it.

mkurz replied:
Maybe I have time to look at this tomorrow?

ioannakok replied:
I merged #26983 and with the help of @kelvin-chappell from @guardian/identity team we no longer pull pre Play 3 dependencies from identity. The only place we still have is `atom-renderer`. I started a branch locally but it looks like I don't have permissions to push so need to sort this out. I'm on leave tomorrow but will have a look on Friday. In the meantime, please feel free to have a look. Locally I have only made 2 changes so far:

1. Changed this [line](https://github.com/guardian/atom-renderer/blob/master/project/plugins.sbt#L1) to `addSbtPlugin("org.playframework.twirl" % "sbt-twirl" % "2.0.4")`
2.  Bumped sbt to 1.9.9

but I'm still getting the old `twirl-api` somehow:

![image](https://github.com/guardian/frontend/assets/19683595/c97dc1f3-4c73-4988-9942-082e4cb80a38)

ioannakok replied:
@mkurz In case you were planning to, please hold off from doing any work on `atom-renderer`. I'm having a look at removing the dependency completely. This is a very old and unmaintained repo.

ioannakok replied:
@mkurz in case you were planning to work on updating `atom-renderer` please hold off. I am working on moving some of the code to frontend and decommissioning the repo.

mkurz replied:
@ioannakok Actually I was trying to upgrade the `atom-renderer` repo locally (before easter) and almost could make it work with some hacks, however I have seen you already opened PRs to drop it, which probably is the better thing to do at this point :+1:

ioannakok replied:
@mkurz Yes, I've just merged the PR: 
* https://github.com/guardian/frontend/pull/27021

I have just rebased. I'm going to raise this PR today - need to add a few things in the description.

mkurz replied:
ah sorry missed this comment, just commented way down below :wink:

ioannakok commented:
> @ioannakok I have rebased against `main` and added a few commits (`sbt` update, `3.0.2` update, update targeting client, etc.) and this now build, but we need to uncomment remaining ignored test and investigate any new failure.

Thanks @mchv! Tests were fixed in the Play 2.9 upgrade:

* https://github.com/guardian/frontend/pull/26783 

I can't see any other ignored tests in this PR, unless I'm missing sth?

mchv commented:
@ioannakok I missed this, sorry! you are right all tests are now included. 
Great work, we seems quite close to be able to ship this 🚀

ioannakok commented:
Moving this to blocked until: 
* https://github.com/guardian/frontend/issues/26997
is completed.

mkurz commented:
@ioannakok I see atom-renderer now got removed and you rebased this branch already. I will take a look at this PR once more later today if you don't mind.

mkurz commented:
@ioannakok You can upgrade scalatestplus-play:
```diff
-  val scalaTestPlus = "org.scalatestplus.play" %% "scalatestplus-play" % "7.0.0" % Test
+  val scalaTestPlus = "org.scalatestplus.play" %% "scalatestplus-play" % "7.0.1" % Test
```

mkurz commented:
Also upgrading [`sbt-buildinfo`](https://github.com/sbt/sbt-buildinfo/releases/tag/v0.12.0) in `project/plugins.sbt` should not hurt:
```diff
-addSbtPlugin("com.eed3si9n" % "sbt-buildinfo" % "0.11.0")
+addSbtPlugin("com.eed3si9n" % "sbt-buildinfo" % "0.12.0")
```

ioannakok commented on facia/test/controllers/front/FrontHeadlineTest.scala:
```diff
@@ -29,8 +29,8 @@ class FrontHeadlineTest extends AnyFunSuite with Matchers {
-    val headline = Helpers.contentAsString(resultFuture)
-    val status = Helpers.status(resultFuture)
+    val headline = Helpers.contentAsString(resultFuture)(timeout)
+    val status = Helpers.status(resultFuture)(timeout)
```
Avoids ambiguous implicit types: https://github.com/guardian/frontend/commit/a1a1e71b0c624001085e55d702e3d2cbc8f638e5

ioannakok commented on project/plugins.sbt:
```diff
@@ -7,11 +7,9 @@ libraryDependencies ++= Seq(
-resolvers ++= Resolver.sonatypeOssRepos("releases") ++ Seq(
-  Classpaths.typesafeReleases,
-)
+resolvers ++= Resolver.sonatypeOssRepos("releases")
```
https://github.com/guardian/frontend/commit/50e0705b9d467b8554438d94dd2489f408e46d9e

ioannakok commented on project/plugins.sbt:
```diff
@@ -21,6 +19,6 @@ addSbtPlugin("com.github.sbt" % "sbt-git" % "2.0.1")
 addSbtPlugin("org.scalameta" % "sbt-scalafmt" % "2.4.0")
 
-addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.10.0-RC1")
+addDependencyTreePlugin
```
https://github.com/guardian/frontend/commit/5b3ed761cb6f74dee0b49c47c8eebb08e9645031

mxdvl commented on common/app/common/TrailsToRss.scala:
```diff
@@ -158,10 +158,6 @@ object TrailsToRss {
       pressedPage.collections
         .filterNot(_.config.excludeFromRss)
         .flatMap(_.curatedPlusBackfillDeduplicated)
-        .filter {
-          case _: LinkSnap => false
-          case _           => true
-        }
```
Could we elaborate on why this has been dropped?

mkurz replied:
For me it seems this has nothing to do with the Play upgrade?

mchv replied:
This is coming from the upgrade of the scala version, which through improvement of types pattern matching can detect that a value of type `model.pressed.PressedContent` can not also be a `com.gu.facia.api.models.LinkSnap`. The filtering code is therefore unnecessary and can be safely removed.

mchv replied:
We can't just keep that code, as it will be considered a compile error.

mxdvl commented on project/ProjectSettings.scala:
```diff
@@ -32,7 +32,7 @@ object ProjectSettings {
     Compile / doc / sources := Seq.empty,
     Compile / doc := target.map(_ / "none").value,
-    scalaVersion := "2.13.12",
+    scalaVersion := "2.13.13",
```
Is it strictly necessary to update the Scala version for this change?

mkurz replied:
It is not absolutly necessary, it could be done in another PR probably, however I recommend to do it in par with the Play upgrade for now, because we already test Play 3.0.2 with Scala 2.13.13, so this should be safe. Same with the sbt upgrade.

In general I recommend you to upgrade to the latest Scala versions as soon as they come out. I have seen that sometimes you lack behind a couple of (patch) versions in various of your repos. I get the "never change a running system" thing, however usually the Scala upgrades should apply safely and sometimes they bring fixes and small new features and also make life easier to upgrade to newer Play (patch) releases later or even JVM (patch) releases.

mchv replied:
I think it is a good practice to upgrade to the latest version of scala environment together, rather than trying to go with atomic PRs, because you benefits from latest bugs fixes.

mchv replied:
I have now crafted this https://github.com/guardian/recommendations/pull/164

mxdvl commented on project/build.properties:
```diff
@@ -1 +1 @@
-sbt.version=1.9.7
+sbt.version=1.9.9
```
Is it strictly necessary to update the SBT version in this PR?

mchv replied:
Yes, it is a good practice to upgrade to latest version of `scala`, `sbt` and sbt plugins when upgrading project.

mxdvl reviewed (approved):
This is a tremendous effort — thanks to everyone involved. 👏

I must ask, however, whether there were a few changes here that are not strictly necessary for the Play 3 upgrade, as we [do recommend PRs remain as small as possible — ideally atomic](https://github.com/guardian/recommendations/blob/main/pull-requests.md#small). It's not a blocker, but it may muddle the waters if anything were to go awry and we'd need to identify the root cause. I'm specifically asking about the changes to SnapLink filtering, SBT version and Scala version.

ioannakok commented:
> This is a tremendous effort — thanks to everyone involved. 👏
> 
> I must ask, however, whether there were a few changes here that are not strictly necessary for the Play 3 upgrade, as we [do recommend PRs remain as small as possible — ideally atomic](https://github.com/guardian/recommendations/blob/main/pull-requests.md#small). It's not a blocker, but it may muddle the waters if anything were to go awry and we'd need to identify the root cause. I'm specifically asking about the changes to SnapLink filtering, SBT version and Scala version.

Thanks @mxdvl for your review and everyone for their contributions! I will split these changes into separate PRs.
